### DISCUSSION TITLE
RCA analog audio to internal ADC to I2S

### DISCUSSION THREAD

**Discussion by StefanoLassandro** [pschatzmann/arduino-audio-tools#1683] [Q&A] [answered]
I'm working with an ESP32 board.

### OBJECTIVE:
The objective is to get an analog audio stream generated by another ESP32, (intended to be used as RCA cable type audio), passing it through the internal ADC, then filter it (as of now I'm not applying any filters, because I first want the thing to work), then output on an I2S bus (which finally passes through a DAC).

### WHAT WORKS:
I've read the frequent issues page.
I've tested the DAC with this example sketch: https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-stream/streams-generator-i2s (I've only modified the I2S pins) and the clean sine wave can be heard clearly.
I've tested the internal ADC with the following sketch:

> #include "AudioTools.h"
> 
> AudioInfo info(44100, 2, 16);
> AnalogAudioStream in;
> CsvOutput<int32_t> csvStream(Serial);
> StreamCopy copier(csvStream, in/*i2sStream*/); // copy analog audio stream to csvStream
> 
> // Arduino Setup
> void setup(void) {
> Serial.begin(115200);
> AudioLogger::instance().begin(Serial, AudioLogger::Info);
>     
> // RX automatically uses port 0 with pin GPIO34
> auto cfgRx = in.defaultConfig(RX_MODE);
> cfgRx.copyFrom(info);
> in.begin(cfgRx);
> 
> csvStream.begin(info);
> }
> 
> // Arduino loop - copy data
> void loop() {
> copier.copy();
> }

And I'm definitely getting some readings from GPIO34:
https://imgur.com/a/GPx32bG (screenshot)

### WHAT DOESN'T WORK
I came up with this sketch, which if I've understood everything correctly should just copy what's coming from the ADC to the DAC.

> #include "AudioTools.h"
> 
> AudioInfo info(44100, 2, 16);
> AnalogAudioStream in; 
> I2SStream out;                        
> StreamCopy copier(out, in); // copy in to out
> 
> // Arduino Setup
> void setup(void) {
>   Serial.begin(115200);
>   //AudioLogger::instance().begin(Serial, AudioLogger::Debug);
> 
>   // RX automatically uses port 0 with pin GPIO34
>   auto cfgRx = in.defaultConfig(RX_MODE);
>   cfgRx.copyFrom(info);
>   in.begin(cfgRx);
>  
>   // TX on I2S_NUM_1 
>   auto cfgTx = out.defaultConfig(TX_MODE);
>   cfgTx.i2s_format = I2S_STD_FORMAT;
>   cfgTx.is_master = true;
>   cfgTx.port_no = 1;
>   cfgTx.pin_ws = 17; // previously 15
>   cfgTx.pin_bck = 18; // previously 14
>   cfgTx.pin_data = 19; // previously 22
>   cfgTx.copyFrom(info);
>   out.begin(cfgTx);
> }
> 
> // Arduino loop - copy data 
> void loop() {
>   copier.copy();
> }

But now I can't hear anything coming out of the DAC.

### OTHER (hopefully useful) INFORMATION

1. With a basic multimeter, I measure a voltage of about 2.6 Volts on the RCA audio that's being emitted by the other ESP32 board, which seems a bit too high seen that the voltage I've found to be right (online) should be 2V peak to peak.
2. Currently there are no components in between the board, just jumper wires.
3. Both ESP32 boards are the same version, the generic ESP32 development module sold by AZ-Delivery (WROOM-32).
4. Power supply is my laptop, but I do have wall adapters and power banks.

Thanks a lot for reading. Any help will be widely appreciated :D

**Comment by pschatzmann** [comment, marked as answer]
Your CSVOutput has a bug: it must be <int16_t>. I also think you have a bug with the number of channels: if you record only on 1 pin you have a mono signal, so the number of channels should be 1 and not 2.

Finally, I don't think you can use the internal DAC for this: but test this yourself by recording a sine wave. The goal is to see a proper 0 centered sine wave in the Arduino Serial Plotter.

I expect that your ADC will only show the + part of the audio and cut off the - part of the sine wave...

**Reply by StefanoLassandro** [reply]
Thanks for your answer.
I've fixed the bugs you mentioned and experimented around a bit, but I still don't get any output from the DAC.
Because I unfortunately don't have an oscilloscope nor a func generator, I was able to test the circuit with a sine wave coming from my laptop, amplifying it a bit, a bit: The input I read from the ESP32 is basically a reversed saw wave with no negative voltages, something like this:
```
_____            _____            __
     \_         |     \_         |
       \________/        \_______/
```

Changing the sine wave frequency does indeed change the frequency of this wave, so I can tell the wave is the one coming from my laptop. Having a function generator outputting directly at the right voltage would be way better, I know :(

Also, why exactly do you say the internal ADC is not good for this? Is it meant to be used just for microphones?

That said, I've ordered an ADC module which seems very flexible, hopefully I'll manage to make it work with this ( https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-stream/streams-i2s-filter-i2s ) example.

Thanks a ton for the help.

**Comment by pschatzmann** [comment]
To center the sine wave (for audio) you can use cfgRx.auto_center_read = true
You can also play with the attenuation db  (e.g. adc_attenuation = ADC_ATTEN_DB_12;) 
)

**Reply by StefanoLassandro** [reply]
Thank you very much,
The auto centering worked flawlessy, but unfortunately (even with attenuation) the waveform is maintaining its weird shape. I've also tried lowering the amplitude from the amplifier I'm using, but it just maintains its shape and finally disappears.
Here's how it looks: https://imgur.com/a/FkTukWS (screenshot)

Also, I still can't hear a thing in the non-debug sketch (the second one in the OP).

Thanks again for answering.

**Reply by pschatzmann** [reply]
This shape does not make any sense. Did you set the log level to Warning to prevent that the logger is messing up the output?
You can also increase the frequency to see shorter waves

**Reply by StefanoLassandro** [reply]
I've commented the line setting the log altogether, so that there was none.
Also, I can unfortunately confirm that changing the frequency of the wave does not change its shape, just its period.
Thanks for answering, hope I'll see better results with the IS2 ADC module that's yet to arrive.
